File: problemas/haskell/FINDSR.hs
import           Data.Array (Array, bounds, listArray, (!))

main :: IO ()
main = interact $ unlines . map parse . takeWhile (/= "*") . lines
  where
    parse :: String -> String
    parse = show . proccess
    proccess :: String -> Int
    proccess xs =
      let pi_array = ptable xs
          m = snd $ bounds pi_array
          last = snd $ pi_array ! m
          root_length = m - last
       in if n `mod` root_length == 0
            then n `div` root_length
            else 1


ptable :: Eq a => [a] -> Array Int (a, Int)
ptable as = xa
  where
    xa = listArray (1, length as) (zip as (scanl op 0 (tail as)))
    op k b
      | a == b = k + 1
      | k == 0 = 0
      | otherwise = op p b
      where
        a = fst (xa ! (k + 1))
        p = snd (xa ! k)

File: problemas/cpp/FINDSR.cpp
#include <iostream>
#include <vector>

using namespace std;

vector<int> compute_prefix_function(const string &pattern) {
  int m = pattern.length();
  vector<int> pi(m);
  pi[0] = 0;
  int k = 0;
  for (int q = 1; q < m; q++) {
    while (k > 0 && pattern[k] != pattern[q])
      k = pi[k - 1];
    if (pattern[k] == pattern[q])
      k++;
    pi[q] = k;
  }
  return pi;
}

int main() {
  string s;
  while (cin >> s && s != "*") {
    vector<int> pi = compute_prefix_function(s);

    int m = s.size();
    int last = pi[m - 1];
    int root_length = m - last;

    if (m % root_length == 0)
      cout << m / root_length << endl;
    else
      cout << 1 << endl;
  }
  return 0;
}
